kernel/src/process/wait.rs

```diff
@@ -31,26 +32,49 @@ bitflags! {
 }
 
 impl WaitOptions {
-    pub fn supported(&self) -> bool {
-        let unsupported_flags = WaitOptions::all() - WaitOptions::WNOHANG;
-        !self.intersects(unsupported_flags)
+    pub fn check(&self) -> Result<()> {
+        if self.intersects(WaitOptions::WSTOPPED | WaitOptions::WCONTINUED)
+            && self.contains(WaitOptions::WNOWAIT)
+        {
+            return_errno_with_message!(
+                Errno::EINVAL,
+                "WNOWAIT cannot be used toghther with WSTOPPED or WCONTINUED"
+            );
+        }
+
+        let supported_args = WaitOptions::WNOHANG
+            | WaitOptions::WSTOPPED
+            | WaitOptions::WCONTINUED
+            | WaitOptions::WNOWAIT;
+        if !supported_args.contains(*self) {
+            warn!(
+                "unsupported wait options are found: {:?}",
+                *self - supported_args
+            );
+        }
+
+        Ok(())
     }
 }
 
-pub fn wait_child_exit(
+pub fn do_wait(
     child_filter: ProcessFilter,
     wait_options: WaitOptions,
     ctx: &Context,
-) -> Result<Option<Arc<Process>>> {
-    let current = ctx.process;
+) -> Result<Option<WaitStatus>> {
+    wait_options.check()?;
+
     let zombie_child = with_sigmask_changed(
         ctx,
         |sigmask| sigmask + SIGCHLD,
         || {
-            current.children_wait_queue().pause_until(|| {
-                let unwaited_children = current
-                    .children()
-                    .lock()
+            ctx.process.children_wait_queue().pause_until(|| {
+                // Acquire the children lock at first to prevent race conditions.
+                // We want to ensure that multiple waiting threads
+                // do not return the same waited process status.
+                let mut children_lock = ctx.process.children().lock();
+
+                let unwaited_children = children_lock
                     .values()
                     .filter(|child| match child_filter {
                         ProcessFilter::Any => true,
```

Comment:
Maybe remove the following `cloned()` as well. It does not seem necessary since we hold the children lock in the entire closure.